### PR TITLE
CP-2658 Refactor platform configuration

### PR DESCRIPTION
**Pull request by evanweible-wf**
_Fixes #125._

## Improvement
Currently, configuring w_transport for a platform (browser, VM, test) is all-or-nothing. This results in some less-than-ideal consequences:

- The platform can only be configured globally. In other words, the configuration is applied to every request and every WebSocket.

- Libraries cannot specify which platform they'd like to use.

  > If a library were to configure w_transport for browser or VM, it _might_ work, but it's up to the consumer at the top of the chain (application) to have the final say.

## Changes

- Added a `TransportPlatform` class. This is essentially the same as the `PlatformAdapter` class, but will now be exported publicly so I took the opportunity to rename it.

  A `TransportPlatform` implementation must know how to construct instances of all the different `BaseRequest`s, the `HttpClient`, and the `WebSocket`.

- Added a `BrowserTransportPlatform` class that provides implementations of the transport classes that work in the browser.

- Added a `browserTransportPlatform` constant that is an instance of the aforementioned `BrowserTransportPlatform` class.

- Added a `BrowserTransportPlatformWithSockJS` class that extends `BrowserTransportPlatform` but overrides the `WebSocket` factory to construct an instance that utilizes SockJS instead of native WebSockets.

- Added a `browserTransportPlatformWithSockJS` constant that is an instance of the aforementioned `BrowserTransportPlatformWithSockJS` with default SockJS settings.

- Added a `VMTransportPlatform` class that provides implementations of the transport classes that work on the Dart VM.

- Added a `vmTransportPlatform` constant that is an instance of the aforementioned `VMTransportPlatform` class.

- All `BaseRequest` classes and the `WebSocket` class now accept an optional `TransportPlatform transportPlatform` parameter. This allows injection of the transport platform to be used for individual instances.

- The `HttpClient` class now also accepts an optional `TransportPlatform transportPlatform` parameter. This transport platform will be injected into each request or WebSocket instance that gets created from that client.

- Added a `globalTransportPlatform` variable that acts as a singleton and allows consumers to set the default, global transport platform. If a transport class is explicitly given a `TransportPlatform` instance upon construction, the `globalTransportPlatform` will be used. Setting this is equivalent to using the `configureWTransportForBrowser()` or `configureWTransportForVM()` functions.

- Added an `install()` method to `MockTransports` - this will serve as the new way to configure w_transport for tests (in place of calling `configureWTransportForTest()`).

## Documentation Updates
I've updated some doc comments in this PR, but I will be making updates to the README/guides to reflect this new configuration pattern in the 3.0.0 docs branch (https://github.com/Workiva/w_transport/pull/210).

## Testing
- [x] CI passes
- [x] Examples behave as expected
- [x] Pull this into truss, update the platform configuration in one of the examples and verify that it behaves as expected

## Code Review

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Raven
Number of Findings: 1
* Watched keyword ```withCredentials``` in ```lib/src/http/mock/request_mixin.dart``` line(s) ['80'] added

As a result of the finding(s) listed above, a security review is required for this pull request.

**Comment by rmconsole-wf** [comment]
@evanweible-wf
* It looks like you have not included any tickets in the title of this PR, but this repo requires tickets for all pull requests. Please create a ticket for this change, or click <a href='https://w-rmconsole.appspot.com/api/v1/pulls/85858577/addTicketToTitle'>here</a> to have Rosie create one for you.

**Comment by evanweible-wf** [comment]
Rebased against latest 3.0.0-wip

**Comment by maxwellpeterson-wf** [comment]
+1

File: lib/src/http/mock/request_mixin.dart [hunk 1]
```diff
@@ -60,6 +67,29 @@ abstract class MockRequestMixin implements MockBaseRequest, CommonRequest {
...
+      ..autoRetry = autoRetry
+      ..headers = headers
+      ..requestInterceptor = requestInterceptor
+      ..responseInterceptor = responseInterceptor
+      ..timeoutThreshold = timeoutThreshold
+      ..uri = uri
+      ..withCredentials = withCredentials;
+
+    // Encoding cannot be set on MultipartRequests
+    if (this is! MultipartRequest) {
```
Comment (dustinlessard-wf) [inline lib/src/http/mock/request_mixin.dart]: I've never seen this before.  I like it!

**Comment by dustinlessard-wf** [comment]
+1

**Comment by evanweible-wf** [comment]
@jayudey-wf ready for you

**Comment by evanweible-wf** [comment]
@maxwellpeterson-wf @dustinlessard-wf Could you review the last commit?

**Comment by maxwellpeterson-wf** [comment]
+1

**Comment by dustinlessard-wf** [comment]
+1

**Comment by evanweible-wf** [comment]
@jayudey-wf ready for you again

**Comment by maxwellpeterson-wf** [comment]
+10
  - CI passes
  - Examples behave as expected
  - Pulled into truss, updated the platform configuration in one of the examples and verifed that it behaves as expected

**Comment by maxwellpeterson-wf** [comment]
QA +1
- [x] Testing instruction
- [x] Dev +1's
- [x] Dev/QA +10 with detail of what was tested
- [x] Unit tests created/updated 
- [x] All unit tests pass

Merging into master.

**Comment by evanweible-wf** [comment]
@rmconsole-wf run_merge_script

**Comment by maxwellpeterson-wf** [comment]
+1 security
- This refactor has no new security implications

**Pull request merged**